sw/device/lib/dif/dif_rv_timer.c

```diff
@@ -0,0 +1,364 @@
+  if (timer == NULL || hart_id >= timer->config.hart_count ||
+      comp_id >= timer->config.comparator_count) {
+    return kDifRvTimerBadArg;
+  }
+
+  ptrdiff_t reg = irq_reg_for_hart(hart_id, timer->config.comparator_count,
+                                   RV_TIMER_INTR_STATE0_REG_OFFSET);
+
+  mmio_region_write32(timer->base_addr, reg, 1 << comp_id);
+
+  return kDifRvTimerOk;
+}
+
+dif_rv_timer_result_t dif_rv_timer_irq_disable(const dif_rv_timer_t *timer,
+                                               uint32_t hart_id,
+                                               uint32_t *state) {
+  if (timer == NULL || hart_id >= timer->config.hart_count) {
+    return kDifRvTimerBadArg;
+  }
+
+  ptrdiff_t reg = irq_reg_for_hart(hart_id, timer->config.comparator_count,
+                                   RV_TIMER_INTR_ENABLE0_REG_OFFSET);
+
+  if (state != NULL) {
+    *state = mmio_region_read32(timer->base_addr, reg);
+  }
+
+  mmio_region_write32(timer->base_addr, reg, 0);
+
+  return kDifRvTimerOk;
+}
+
+dif_rv_timer_result_t dif_rv_timer_irq_restore(const dif_rv_timer_t *timer,
+                                               uint32_t hart_id,
+                                               uint32_t state) {
+  if (timer == NULL || hart_id >= timer->config.hart_count) {
+    return kDifRvTimerBadArg;
+  }
+
+  ptrdiff_t reg = irq_reg_for_hart(hart_id, timer->config.comparator_count,
+                                   RV_TIMER_INTR_ENABLE0_REG_OFFSET);
+
+  mmio_region_write32(timer->base_addr, reg, state);
+
+  return kDifRvTimerOk;
+}
+
+dif_rv_timer_result_t dif_rv_timer_irq_force(const dif_rv_timer_t *timer,
+                                             uint32_t hart_id,
+                                             uint32_t comp_id) {
+  if (timer == NULL || hart_id >= timer->config.hart_count ||
+      comp_id >= timer->config.comparator_count) {
+    return kDifRvTimerBadArg;
+  }
+
+  ptrdiff_t reg = irq_reg_for_hart(hart_id, timer->config.comparator_count,
+                                   RV_TIMER_INTR_TEST0_REG_OFFSET);
+
+  mmio_region_write32(timer->base_addr, reg, 1 << comp_id);
+
+  return kDifRvTimerOk;
+}
+
+dif_rv_timer_result_t dif_rv_timer_reset(const dif_rv_timer_t *timer) {
+  if (timer == NULL) {
+    return kDifRvTimerBadArg;
+  }
+
+  // Disable all counters.
+  mmio_region_write32(timer->base_addr, RV_TIMER_CTRL_REG_OFFSET, 0x0);
+
+  for (uint32_t hart_id = 0; hart_id < timer->config.hart_count; ++hart_id) {
+    // Clear and disable all interrupts.
+    ptrdiff_t irq_status =
+        irq_reg_for_hart(hart_id, timer->config.comparator_count,
+                         RV_TIMER_INTR_STATE0_REG_OFFSET);
+    ptrdiff_t irq_enable =
+        irq_reg_for_hart(hart_id, timer->config.comparator_count,
+                         RV_TIMER_INTR_ENABLE0_REG_OFFSET);
+    mmio_region_write32(timer->base_addr, irq_status, UINT32_MAX);
+    mmio_region_write32(timer->base_addr, irq_enable, 0x0);
```

Comment:
Probably we should clear the interrupts after disabling them.

Reply:
Ooh, good catch!